src/controllers/controller.h

```diff
@@ -169,12 +176,32 @@ class Controller : public QObject, ConstControllerPresetVisitor {
     bool m_bLearning;
     QElapsedTimer m_userActivityInhibitTimer;
 
-    UserSettingsPointer m_pConfig;
-
+    friend class ControllerJSProxy;
     // accesses lots of our stuff, but in the same thread
     friend class ControllerManager;
     // For testing
     friend class ControllerPresetValidationTest;
 };
 
+// An object of this class gets exposed to the JS engine, so the methods of this class
+// constitute the api that is provided to scripts under "controller" object.
+// See comments on ControllerEngineJSProxy.
+class ControllerJSProxy : public QObject {
+    Q_OBJECT
+  public:
+    ControllerJSProxy(Controller* m_pController)
+            : m_pController(m_pController) {
+    }
+
+    // The length parameter is here for backwards compatibility for when scripts
+    // were required to specify it.
+    Q_INVOKABLE virtual void send(QList<int> data, unsigned int length = 0) {
+        Q_UNUSED(length);
+        m_pController->send(data, data.length());
+    }
+
+  private:
+    Controller* m_pController;
```

Comment:
I guess we could use `Controller* const m_pController;` and thereby prevent any copy or move assignments.